src/components/Match/VisionPage.jsx

```diff
@@ -0,0 +1,231 @@
+       if (tick <= props.value)
+         cls.push(styles['active']);
+
+       return <a key={tick} onClick={() => props.onTickClick(tick)} className={cls.join(' ')} style={{left: percent + '%'}}>{formatSeconds(tick)}</a>
+     })}
+  </div>
+);
+
+
+class PlayerFilter extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.getObserverCount = () => this.props.player.obs_log.length;
+    this.getSentryCount = () => this.props.player.sen_log.length;
+  }
+  
+  generateFilterKey(type) {
+    return `${this.props.player.player_slot}-${type}`;
+  }
+  
+  getMuiThemeProps() {
+    return {
+      fullWidth: true,
+      backgroundColor: _.sample([filterOn, filterOff]),
+      disabledBackgroundColor: filterOff,
+    };
+  }
+
+  render() {
+    const {
+      player,
+      onFilterClick,
+    } = this.props;
+    const obs_count = this.getObserverCount();
+    const sen_count = this.getSentryCount();
+    const [opacityOn, opacityOff] = [1, .4];
+    return (
+      <Row className={styles['filter-row']}
+           middle="xs"
+           between="xs">
+        <Col xs={12} sm={7}>
+          <Row xs>
+            <Col xs={1}></Col>
+            <Col>{heroTd(player)}</Col>
+          </Row>
+        </Col>
+        <Col xs={12} sm={5}>
+          <Row>
+            <Col xs>
+              <Button {...this.getMuiThemeProps()}
+                      label={obs_count}
+                      disabled={obs_count == 0}
+                      backgroundColor={this.generateFilterKey("observer") in this.props.activeFilters ? filterOff : filterOn}
+                      style={{opacity: obs_count > 0 ? opacityOn : opacityOff}}
+                      onClick={() => onFilterClick(this.generateFilterKey("observer"), player.player_slot, "observer")}
+                      icon={<Avatar size={24} src="http://a19a1164.ngrok.io/apps/dota2/images/items/ward_observer_lg.png" />}
+              />
+            </Col>
+            <Col xs>
+              <Button {...this.getMuiThemeProps()}
+                      label={sen_count}
+                      disabled={sen_count == 0}
+                      backgroundColor={this.generateFilterKey("sentry") in this.props.activeFilters ? filterOff : filterOn}
+                      style={{opacity: sen_count > 0 ? opacityOn : opacityOff}}
+                      onClick={() => onFilterClick(this.generateFilterKey("sentry"), player.player_slot, "sentry")}
+                      icon={<Avatar size={24} src="http://a19a1164.ngrok.io/apps/dota2/images/items/ward_sentry_lg.png" />}
+              />
+            </Col>
+          </Row>
+        </Col>
+      </Row>
+    );
+  }
+}
+
+const PlayersFilter = ({ activeFilters, players, onFilterClick }) => (
+  <Paper className={styles['ward-log-player-filter']}>
+    {players.map((p,i) => <PlayerFilter player={p} activeFilters={activeFilters} onFilterClick={onFilterClick} />)}
+  </Paper>
+);
+
+import Immutable from 'seamless-immutable';
```

Comment:
Yes. You can compare deeply nested objects/arrays by reference.
I need to extract this though (the whole ward_log should be Immutable, i.e. in renderMatch)